# University Marks Sheet EDA 


In [147]:

import os
for dirname, _, filenames in os.walk('/kaggle/input/university-students-marks-sheet/'):
    for filename in filenames:
        print(os.path.join(dirname, filename))



/kaggle/input/university-students-marks-sheet/Univeristy_Results.xls


In [148]:
import numpy as np # linear algebra
import pandas as pd 

<h5>  Was having trouble find actual file path was a bit confused with excel files in a directory skip this section <h5>

In [149]:
!pip install xlrd

In [150]:
import os

In [151]:
# directory containing the Excel files
directory = '/kaggle/input/university-students-marks-sheet'

In [152]:
# List all files in the directory
files = os.listdir(directory)

In [153]:
# Filter Excel files
excel_files = [file for file in files if file.endswith('.xlsx') or file.endswith('.xls')]

In [154]:
# Initialize an empty list to store DataFrames
dfs = []

In [155]:
# Loop through each Excel file, read it into a DataFrame, and append to the list
for file in excel_files:
    file_path = os.path.join(directory, file)
    df = pd.read_excel(file_path)
    dfs.append(df)

In [156]:
# Concatenate all DataFrames into a single DataFrame
combined_df = pd.concat(dfs, ignore_index=False)

In [157]:
combined_df

Table Unnamed: 1 Unnamed: 2  Unnamed: 3          Unnamed: 4
0      Marks          GP      Grade         NaN                 NaN
1           -          0          F      -0.003     Zero point zero
2           0          0          F      -0.003     Zero point zero
3          *1          0          F      -0.002     Zero point zero
4          *2          0          F      -0.001     Zero point zero
...       ...        ...        ...         ...                 ...
55002     NaN        NaN        NaN      99.960  Hundred point zero
55003     NaN        NaN        NaN      99.970  Hundred point zero
55004     NaN        NaN        NaN      99.980  Hundred point zero
55005     NaN        NaN        NaN      99.990  Hundred point zero
55006     NaN        NaN        NaN     100.000  Hundred point zero

[55007 rows x 5 columns]

Still cound not sort out the problem!!!!!

# Loading Datasets Files

In [158]:
import pandas as pd

# Read all sheets into a dictionary of DataFrames
dfs = pd.read_excel('/kaggle/input/university-students-marks-sheet/Univeristy_Results.xls', sheet_name=None)

# Access each DataFrame by sheet name
for sheet_name, df in dfs.items():
    print(f"Sheet Name: {sheet_name}")
    print(df)


Sheet Name: Table
        Table Unnamed: 1 Unnamed: 2  Unnamed: 3          Unnamed: 4
0      Marks          GP      Grade         NaN                 NaN
1           -          0          F      -0.003     Zero point zero
2           0          0          F      -0.003     Zero point zero
3          *1          0          F      -0.002     Zero point zero
4          *2          0          F      -0.001     Zero point zero
...       ...        ...        ...         ...                 ...
55002     NaN        NaN        NaN      99.960  Hundred point zero
55003     NaN        NaN        NaN      99.970  Hundred point zero
55004     NaN        NaN        NaN      99.980  Hundred point zero
55005     NaN        NaN        NaN      99.990  Hundred point zero
55006     NaN        NaN        NaN     100.000  Hundred point zero

[55007 rows x 5 columns]
Sheet Name: Sessional
                                Roll No           Name Quiz 15   \
0                                   NaN            

# Loading Sessional File from Dataset Seperatly

In [159]:
# Read specific sheets into separate DataFrames
sessional_df = pd.read_excel('/kaggle/input/university-students-marks-sheet/Univeristy_Results.xls', sheet_name='Sessional')

In [160]:
sessional_df

Roll No           Name Quiz 15   \
0                                   NaN            NaN  Quiz 01   
1                                     1  Ashir Mehfooz       14   
2                                     2    Atif Raftad        4   
3                                     3     Saiqa Aziz       15   
4                                     8   Ozair Minhas        6   
..                                  ...            ...      ...   
220      Course Instructor Signature:              NaN      NaN   
221                                 NaN            NaN      NaN   
222                                 NaN            NaN      NaN   
223                                   1            NaN        2   
224  Departmental Examination Committee            NaN      NaN   

                  Unnamed: 3 Unnamed: 4 Assignments 15  Unnamed: 6 Unnamed: 7  \
0                    Quiz 02   AVG (15)       Assig. 01   Assig.02   AVG (15)   
1                         14         14              13         13         13   
2                         10          7               4          5        4.5   
3                         11         13              14         13       13.5   
4                          5        5.5               4          6          5   
..                       ...        ...             ...        ...        ...   
220  Chairman/HoD Signature:        NaN             NaN        NaN        NaN   
221                      NaN        NaN             NaN        NaN        NaN   
222                      NaN        NaN             NaN        NaN        NaN   
223                      NaN        NaN             NaN        NaN        NaN   
224                      NaN        NaN             NaN        NaN        NaN   

    Mid Term    Q+A+Mid  
0         45   Total 75  
1         41         68  
2         30       41.5  
3         34       60.5  
4         12       22.5  
..       ...        ...  
220      NaN        NaN  
221      NaN        NaN  
222      NaN        NaN  
223      NaN        NaN  
224      NaN        NaN  

[225 rows x 10 columns]

# Loading Final sheet from Dataset

In [161]:

final_df = pd.read_excel('/kaggle/input/university-students-marks-sheet/Univeristy_Results.xls', sheet_name='Final')

In [162]:
final_df

Roll No. Name of Student Total Marks (150)  \
0                                   NaN             NaN    Sessional (75)   
1                                     1   Ashir Mehfooz                68   
2                                     2     Atif Raftad              41.5   
3                                     3      Saiqa Aziz              60.5   
4                                     8    Ozair Minhas              22.5   
..                                  ...             ...               ...   
230                                 NaN             NaN               NaN   
231                                 NaN             NaN               NaN   
232                                 NaN             NaN               NaN   
233                                   1             NaN                 2   
234  Departmental Examination Committee             NaN               NaN   

        Unnamed: 3 Unnamed: 4                         Marks in Words  \
0    Terminal (75)  Total 150                                    NaN   
1               55        123        One Hundred & Twenty Three Only   
2               45       86.5              Eighty Six Pont Five Zero   
3               58      118.5  One Hundred & Eighteen Pont Five Zero   
4               44       66.5               Sixty Six Pont Five Zero   
..             ...        ...                                    ...   
230            NaN        NaN                                    NaN   
231            NaN        NaN                                    NaN   
232            NaN        NaN                                    NaN   
233            NaN        NaN                                      3   
234            NaN        NaN                                    NaN   

             % Letter Grade   GP  
0          NaN          NaN  NaN  
1    82.000000            A  4.0  
2    57.666667            C  2.5  
3    79.000000           B+  3.9  
4    44.333333            F  0.0  
..         ...          ...  ...  
230        NaN          NaN  NaN  
231        NaN          NaN  NaN  
232        NaN          NaN  NaN  
233        NaN          NaN  NaN  
234        NaN          NaN  NaN  

[235 rows x 9 columns]

# Check Version 2 for Data Cleaning Process

In [163]:
sessional_df.columns

Index(['Roll No', 'Name', 'Quiz 15 ', 'Unnamed: 3', 'Unnamed: 4',
       'Assignments 15 ', 'Unnamed: 6', 'Unnamed: 7', 'Mid Term', 'Q+A+Mid'],
      dtype='object')

In [164]:
sessional_df.shape

(225, 10)

In [165]:
final_df.columns

Index(['Roll No.', 'Name of Student', 'Total Marks (150)', 'Unnamed: 3',
       'Unnamed: 4', 'Marks in Words', '%', 'Letter Grade', 'GP'],
      dtype='object')

In [166]:
final_df.shape

(235, 9)

In [167]:
# Selecting desired columns from sessional_df
sessional_selected = sessional_df[['Roll No', 'Name', 'Unnamed: 4', 'Unnamed: 7', 'Mid Term','Q+A+Mid']]

# Selecting desired columns from final_df and renaming them for consistency
final_selected = final_df[['Unnamed: 3', 'Unnamed: 4','Marks in Words','%', 'Letter Grade', 'GP']]

# Concatenating the DataFrames
combined_df = pd.concat([sessional_selected, final_selected], axis=1)

# Displaying the combined DataFrame
combined_df


Roll No           Name Unnamed: 4 Unnamed: 7 Mid Term    Q+A+Mid  \
0       NaN            NaN   AVG (15)   AVG (15)       45   Total 75   
1         1  Ashir Mehfooz         14         13       41         68   
2         2    Atif Raftad          7        4.5       30       41.5   
3         3     Saiqa Aziz         13       13.5       34       60.5   
4         8   Ozair Minhas        5.5          5       12       22.5   
..      ...            ...        ...        ...      ...        ...   
230     NaN            NaN        NaN        NaN      NaN        NaN   
231     NaN            NaN        NaN        NaN      NaN        NaN   
232     NaN            NaN        NaN        NaN      NaN        NaN   
233     NaN            NaN        NaN        NaN      NaN        NaN   
234     NaN            NaN        NaN        NaN      NaN        NaN   

        Unnamed: 3 Unnamed: 4                         Marks in Words  \
0    Terminal (75)  Total 150                                    NaN   
1               55        123        One Hundred & Twenty Three Only   
2               45       86.5              Eighty Six Pont Five Zero   
3               58      118.5  One Hundred & Eighteen Pont Five Zero   
4               44       66.5               Sixty Six Pont Five Zero   
..             ...        ...                                    ...   
230            NaN        NaN                                    NaN   
231            NaN        NaN                                    NaN   
232            NaN        NaN                                    NaN   
233            NaN        NaN                                      3   
234            NaN        NaN                                    NaN   

             % Letter Grade   GP  
0          NaN          NaN  NaN  
1    82.000000            A  4.0  
2    57.666667            C  2.5  
3    79.000000           B+  3.9  
4    44.333333            F  0.0  
..         ...          ...  ...  
230        NaN          NaN  NaN  
231        NaN          NaN  NaN  
232        NaN          NaN  NaN  
233        NaN          NaN  NaN  
234        NaN          NaN  NaN  

[235 rows x 12 columns]

In [168]:
combined_df.columns

Index(['Roll No', 'Name', 'Unnamed: 4', 'Unnamed: 7', 'Mid Term', 'Q+A+Mid',
       'Unnamed: 3', 'Unnamed: 4', 'Marks in Words', '%', 'Letter Grade',
       'GP'],
      dtype='object')

In [169]:
# Changing combined_df columns name
new_columns = {'Roll No': 'Student_id',
               'Unnamed: 4':'Quiz_Marks',
               'Unnamed: 7':'Assignment_Marks',
               'Q+A+Mid':'Sessional(75)',
               'Unnamed: 3':'Terminal(75)',
               'Unnamed: 4':'Obtained(150)',
               'Marks in Words':'Marks_in_words',
               '%':'Percentage',
               'Letter Grade':'Grade'
               ,'GP':'GPA'
              }

# Dataframe with New columbs Name

In [171]:
#it will not showing output cuz changes are made permanently 
combined_df.rename(columns=new_columns, inplace=True)

In [172]:
# dropping the first row at location 0
combined_df.drop(0, inplace=True)

In [173]:
combined_df.tail(20)

Student_id   Name Obtained(150)  \
215                                                NaN  Fırat             4   
216                                                222  Gamze          11.5   
217  Abbreviations: Q=Quiz, A=Assignment, Att=Atten...    NaN           NaN   
218                                                NaN    NaN           NaN   
219                                                NaN    NaN           NaN   
220                     Course Instructor Signature:      NaN           NaN   
221                                                NaN    NaN           NaN   
222                                                NaN    NaN           NaN   
223                                                  1    NaN           NaN   
224                 Departmental Examination Committee    NaN           NaN   
225                                                NaN    NaN           NaN   
226                                                NaN    NaN           NaN   
227                                                NaN    NaN           NaN   
228                                                NaN    NaN           NaN   
229                                                NaN    NaN           NaN   
230                                                NaN    NaN           NaN   
231                                                NaN    NaN           NaN   
232                                                NaN    NaN           NaN   
233                                                NaN    NaN           NaN   
234                                                NaN    NaN           NaN   

    Assignment_Marks Mid Term Sessional(75) Terminal(75)  \
215               12       27            43          NaN   
216               15       33          59.5           47   
217              NaN      NaN           NaN          NaN   
218              NaN      NaN           NaN          NaN   
219              NaN      NaN           NaN          NaN   
220              NaN      NaN           NaN       Grades   
221              NaN      NaN           NaN           A+   
222              NaN      NaN           NaN            A   
223              NaN      NaN           NaN           B+   
224              NaN      NaN           NaN            B   
225              NaN      NaN           NaN            C   
226              NaN      NaN           NaN            F   
227              NaN      NaN           NaN          NaN   
228              NaN      NaN           NaN          NaN   
229              NaN      NaN           NaN          NaN   
230              NaN      NaN           NaN          NaN   
231              NaN      NaN           NaN          NaN   
232              NaN      NaN           NaN          NaN   
233              NaN      NaN           NaN          NaN   
234              NaN      NaN           NaN          NaN   

               Obtained(150)                      Marks_in_words  Percentage  \
215                       43                   Fourty Three Only   28.666667   
216                    106.5    One Hundred & Six Pont Five Zero   71.000000   
217                      NaN                                 NaN         NaN   
218                      NaN                                 NaN         NaN   
219                      NaN                                 NaN         NaN   
220                      NaN                                 NaN         NaN   
221                        0            A+         90% and above         NaN   
222                       19   A           80% and less than 90%         NaN   
223                       24    B+        72% and less than 80%          NaN   
224                       25  B           65 % and less than 72%         NaN   
225                       54  C           50 % and less than 65%         NaN   
226                       83           F           Less than 50%         NaN   
227                      NaN                                 NaN         NaN   
228

<h4>From index 217 to 234 these rows are not significant for now so im gonna drop them! <h4>

In [174]:
# Dropping multiple rows using range function
combined_df.drop(range(217,235),inplace=True)

<h3> Now my Dataset is how i wanted it to be for further detailed Analysis i have changed columns name for my ease<h3>

In [175]:
combined_df

Student_id             Name Obtained(150) Assignment_Marks Mid Term  \
1            1    Ashir Mehfooz            14               13       41   
2            2      Atif Raftad             7              4.5       30   
3            3       Saiqa Aziz            13             13.5       34   
4            8     Ozair Minhas           5.5                5       12   
5            9  Naveera Subhani             8              4.5       34   
..         ...              ...           ...              ...      ...   
212        218            Yelda            11               13       24   
213        219            Ertan            14                9       11   
214        220           Cansın            13               13       21   
215        NaN            Fırat             4               12       27   
216        222            Gamze          11.5               15       33   

    Sessional(75) Terminal(75) Obtained(150)  \
1              68           55           123   
2            41.5           45          86.5   
3            60.5           58         118.5   
4            22.5           44          66.5   
5            46.5           35          81.5   
..            ...          ...           ...   
212            48           28            76   
213            34           40            74   
214            47           30            77   
215            43          NaN            43   
216          59.5           47         106.5   

                            Marks_in_words  Percentage Grade  GPA  
1          One Hundred & Twenty Three Only   82.000000     A  4.0  
2                Eighty Six Pont Five Zero   57.666667     C  2.5  
3    One Hundred & Eighteen Pont Five Zero   79.000000    B+  3.9  
4                 Sixty Six Pont Five Zero   44.333333     F  0.0  
5                Eighty One Pont Five Zero   54.333333     C  2.2  
..                                     ...         ...   ...  ...  
212                       Seventy Six Only   50.666667     C  2.0  
213                      Seventy Four Only   49.333333     F  0.0  
214                     Seventy Seven Only   51.333333     C  2.0  
215                      Fourty Three Only   28.666667     F  0.0  
216       One Hundred & Six Pont Five Zero   71.000000     B  3.4  

[216 rows x 12 columns]

<h4> Checking a single row from dataset<h4>

In [176]:
#Location 1
combined_df.iloc[0]

Student_id                                        1
Name                                  Ashir Mehfooz
Obtained(150)                                    14
Assignment_Marks                                 13
Mid Term                                         41
Sessional(75)                                    68
Terminal(75)                                     55
Obtained(150)                                   123
Marks_in_words      One Hundred & Twenty Three Only
Percentage                                     82.0
Grade                                             A
GPA                                             4.0
Name: 1, dtype: object

In [177]:
#Location Last
combined_df.iloc[-1]

Student_id                                       222
Name                                           Gamze
Obtained(150)                                   11.5
Assignment_Marks                                  15
Mid Term                                          33
Sessional(75)                                   59.5
Terminal(75)                                      47
Obtained(150)                                  106.5
Marks_in_words      One Hundred & Six Pont Five Zero
Percentage                                      71.0
Grade                                              B
GPA                                              3.4
Name: 216, dtype: object

In [178]:
#In a range of locations
combined_df.iloc[range(111,118)]

Student_id    Name Obtained(150) Assignment_Marks Mid Term Sessional(75)  \
112        118   Defne           2.5                4       34          40.5   
113              Kadir          13.5                5       25          43.5   
114        120    Aslı           7.5                6       23          36.5   
115        121  Ferhat            15                0       23            38   
116        122   Duygu           7.5                3       34          44.5   
117        123   İlker            12               14       24            50   
118        124    İnci           9.5                4       11          24.5   

    Terminal(75) Obtained(150)              Marks_in_words  Percentage Grade  \
112           56          96.5    Ninty Six Pont Five Zero   64.333333     C   
113           22          65.5   Sixty Five Pont Five Zero   43.666667     F   
114            3          39.5  Thirty Nine Pont Five Zero   26.333333     F   
115           60            98            Ninty Eight Only   65.333333     B   
116            7          51.5    Fifty One Pont Five Zero   34.333333     F   
117           58           108    One Hundred & Eight Only   72.000000    B+   
118           32          56.5    Fifty Six Pont Five Zero   37.666667     F   

     GPA  
112  2.9  
113  0.0  
114  0.0  
115  3.0  
116  0.0  
117  3.4  
118  0.0

In [179]:
combined_df.isnull().sum()

Student_id           1
Name                 0
Obtained(150)        0
Assignment_Marks     0
Mid Term             0
Sessional(75)       11
Terminal(75)         3
Obtained(150)       11
Marks_in_words      11
Percentage          11
Grade               11
GPA                 11
dtype: int64

# Finding Index of missing student id

In [180]:
# Find the indices of rows where 'Student_id' is null
null_student_id_indices = combined_df[combined_df['Student_id'].isnull()].index

# Print the indices
print(null_student_id_indices)

Index([215], dtype='int64')


In [181]:
combined_df.iloc[214]

Student_id                        NaN
Name                            Fırat
Obtained(150)                       4
Assignment_Marks                   12
Mid Term                           27
Sessional(75)                      43
Terminal(75)                      NaN
Obtained(150)                      43
Marks_in_words      Fourty Three Only
Percentage                  28.666667
Grade                               F
GPA                               0.0
Name: 215, dtype: object

# Filling missing student id

In [182]:
# Fill null values in the 'Student_id' column with a specified value (e.g., 0)
combined_df['Student_id'].fillna(214, inplace=True)


/tmp/ipykernel_33/1361607406.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  combined_df['Student_id'].fillna(214, inplace=True)


In [183]:
combined_df

Student_id             Name Obtained(150) Assignment_Marks Mid Term  \
1            1    Ashir Mehfooz            14               13       41   
2            2      Atif Raftad             7              4.5       30   
3            3       Saiqa Aziz            13             13.5       34   
4            8     Ozair Minhas           5.5                5       12   
5            9  Naveera Subhani             8              4.5       34   
..         ...              ...           ...              ...      ...   
212        218            Yelda            11               13       24   
213        219            Ertan            14                9       11   
214        220           Cansın            13               13       21   
215        214            Fırat             4               12       27   
216        222            Gamze          11.5               15       33   

    Sessional(75) Terminal(75) Obtained(150)  \
1              68           55           123   
2            41.5           45          86.5   
3            60.5           58         118.5   
4            22.5           44          66.5   
5            46.5           35          81.5   
..            ...          ...           ...   
212            48           28            76   
213            34           40            74   
214            47           30            77   
215            43          NaN            43   
216          59.5           47         106.5   

                            Marks_in_words  Percentage Grade  GPA  
1          One Hundred & Twenty Three Only   82.000000     A  4.0  
2                Eighty Six Pont Five Zero   57.666667     C  2.5  
3    One Hundred & Eighteen Pont Five Zero   79.000000    B+  3.9  
4                 Sixty Six Pont Five Zero   44.333333     F  0.0  
5                Eighty One Pont Five Zero   54.333333     C  2.2  
..                                     ...         ...   ...  ...  
212                       Seventy Six Only   50.666667     C  2.0  
213                      Seventy Four Only   49.333333     F  0.0  
214                     Seventy Seven Only   51.333333     C  2.0  
215                      Fourty Three Only   28.666667     F  0.0  
216       One Hundred & Six Pont Five Zero   71.000000     B  3.4  

[216 rows x 12 columns]

# set student id as index

In [184]:
combined_df.set_index('Student_id', inplace=False)

Name Obtained(150) Assignment_Marks Mid Term  \
Student_id                                                            
1             Ashir Mehfooz            14               13       41   
2               Atif Raftad             7              4.5       30   
3                Saiqa Aziz            13             13.5       34   
8              Ozair Minhas           5.5                5       12   
9           Naveera Subhani             8              4.5       34   
...                     ...           ...              ...      ...   
218                   Yelda            11               13       24   
219                   Ertan            14                9       11   
220                  Cansın            13               13       21   
214                   Fırat             4               12       27   
222                   Gamze          11.5               15       33   

           Sessional(75) Terminal(75) Obtained(150)  \
Student_id                                            
1                     68           55           123   
2                   41.5           45          86.5   
3                   60.5           58         118.5   
8                   22.5           44          66.5   
9                   46.5           35          81.5   
...                  ...          ...           ...   
218                   48           28            76   
219                   34           40            74   
220                   47           30            77   
214                   43          NaN            43   
222                 59.5           47         106.5   

                                   Marks_in_words  Percentage Grade  GPA  
Student_id                                                                
1                 One Hundred & Twenty Three Only   82.000000     A  4.0  
2                       Eighty Six Pont Five Zero   57.666667     C  2.5  
3           One Hundred & Eighteen Pont Five Zero   79.000000    B+  3.9  
8                        Sixty Six Pont Five Zero   44.333333     F  0.0  
9                       Eighty One Pont Five Zero   54.333333     C  2.2  
...                                           ...         ...   ...  ...  
218                              Seventy Six Only   50.666667     C  2.0  
219                             Seventy Four Only   49.333333     F  0.0  
220                            Seventy Seven Only   51.333333     C  2.0  
214                             Fourty Three Only   28.666667     F  0.0  
222              One Hundred & Six Pont Five Zero   71.000000     B  3.4  

[216 rows x 11 columns]

In [185]:
colum_name_change = {'Mid Term':'Mids'}
combined_df.rename(columns=colum_name_change,inplace=True)

In [186]:
combined_df

Student_id             Name Obtained(150) Assignment_Marks Mids  \
1            1    Ashir Mehfooz            14               13   41   
2            2      Atif Raftad             7              4.5   30   
3            3       Saiqa Aziz            13             13.5   34   
4            8     Ozair Minhas           5.5                5   12   
5            9  Naveera Subhani             8              4.5   34   
..         ...              ...           ...              ...  ...   
212        218            Yelda            11               13   24   
213        219            Ertan            14                9   11   
214        220           Cansın            13               13   21   
215        214            Fırat             4               12   27   
216        222            Gamze          11.5               15   33   

    Sessional(75) Terminal(75) Obtained(150)  \
1              68           55           123   
2            41.5           45          86.5   
3            60.5           58         118.5   
4            22.5           44          66.5   
5            46.5           35          81.5   
..            ...          ...           ...   
212            48           28            76   
213            34           40            74   
214            47           30            77   
215            43          NaN            43   
216          59.5           47         106.5   

                            Marks_in_words  Percentage Grade  GPA  
1          One Hundred & Twenty Three Only   82.000000     A  4.0  
2                Eighty Six Pont Five Zero   57.666667     C  2.5  
3    One Hundred & Eighteen Pont Five Zero   79.000000    B+  3.9  
4                 Sixty Six Pont Five Zero   44.333333     F  0.0  
5                Eighty One Pont Five Zero   54.333333     C  2.2  
..                                     ...         ...   ...  ...  
212                       Seventy Six Only   50.666667     C  2.0  
213                      Seventy Four Only   49.333333     F  0.0  
214                     Seventy Seven Only   51.333333     C  2.0  
215                      Fourty Three Only   28.666667     F  0.0  
216       One Hundred & Six Pont Five Zero   71.000000     B  3.4  

[216 rows x 12 columns]

In [187]:
combined_df.columns

Index(['Student_id', 'Name', 'Obtained(150)', 'Assignment_Marks', 'Mids',
       'Sessional(75)', 'Terminal(75)', 'Obtained(150)', 'Marks_in_words',
       'Percentage', 'Grade', 'GPA'],
      dtype='object')

# Filling other columns values "Numerical Columns"

Filling missing values with 0 in Numeric columns

In [188]:
# Fill missing values in numerical columns with the mean
numerical_columns = ['Sessional(75)', 'Terminal(75)']
combined_df[numerical_columns] = combined_df[numerical_columns].fillna(0,inplace=False)

/tmp/ipykernel_33/3291838161.py:3: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  combined_df[numerical_columns] = combined_df[numerical_columns].fillna(0,inplace=False)


Filling with Mean

In [ ]:
# Specify the numerical columns
numerical_columns = ['Sessional(75)', 'Terminal(75)']

# Convert columns to numeric data types
combined_df[numerical_columns] = combined_df[numerical_columns].apply(pd.to_numeric, errors='coerce')

# Fill missing values in numerical columns with the mean
combined_df[numerical_columns] = combined_df[numerical_columns].fillna(combined_df[numerical_columns].mean(),inplace=False)


In [189]:
combined_df.isnull().sum()

Student_id           0
Name                 0
Obtained(150)        0
Assignment_Marks     0
Mids                 0
Sessional(75)        0
Terminal(75)         0
Obtained(150)       11
Marks_in_words      11
Percentage          11
Grade               11
GPA                 11
dtype: int64

In [ ]:
combined_df

<span style="color:red">Currently, I'm excluding the "Marks in words," "Grade," and "GPA" columns from the data filling process as they contain non-numeric data. Utilizing the mode to fill these columns isn't suitable because grades are categorical values and need more sophisticated techniques. Hence, I'll employ machine learning methods such as classification and regression to predict these categorical values based on a set of input features. This approach ensures accurate imputation of grades while leveraging the available dataset effectively.</span>
